internal/addonsdk/sdk.go

```diff
@@ -0,0 +1,215 @@
+package addonsdk
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	addonsv1alpha1 "github.com/openshift/addon-operator/apis/addons/v1alpha1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+var (
+	statusReporterSingleton      *StatusReporter
+	statusReporterSingletonMutex = &sync.Mutex{}
+)
+
+// a nice heartbeat-reporter implemented by the MT-SRE which our tenants can use
+// if they don't like it, they can implement their own heartbeat reporter by creating a type which implements the `addonsdk.statusReporterClient` interface
+type StatusReporter struct {
+	// object provided by the client/tenants which implements the addonsdk.client interface
+	addonInstanceInteractor client
+	addonName               string
+	addonTargetNamespace    string
+
+	// the latest conditions which the heartbeat reporter would be reporting periodically
+	latestConditions []metav1.Condition
+
+	// to control the rate at which the heartbeat reporter would run
+	interval time.Duration
+	ticker   *time.Ticker
+
+	// for effectively communicating the stop and update signals
+	stopperCh chan bool
+	updateCh  chan updateOptions
+
+	// for concurrency-safely executing one instance of heartbeat reporter loop
+	executeOnce sync.Once
+
+	//for tracking if the heartbeat reporter is running or done running
+	doneCh chan bool
+
+	log logr.Logger
+}
+
+// ensure that the `StatusReporter` implements the `addonsdk.statusReporterClient` interface
+var _ statusReporterClient = (*StatusReporter)(nil)
+
+// InitializeStatusReporterSingleton sets up a singleton of the type `StatusReporter` (only if it doesn't exist yet) and returns it to the caller.
+func InitializeStatusReporterSingleton(addonInstanceInteractor client, addonName string, addonTargetNamespace string, logger logr.Logger) *StatusReporter {
+	if statusReporterSingleton == nil {
+		statusReporterSingletonMutex.Lock()
+		defer statusReporterSingletonMutex.Unlock()
```

Comment:
Right but here, the first `if` is to just provide non-locked reads to any future readers/goroutines calling this function because there's no point to Lock/Unlock once the Singleton is initialised as it's gonna be just GET THE SINGLETON after the Singleton is initialized. Hence, the first `if` isn't preceded with Locking it, 
But in theory, before the Singleton is initialized, there can be two (or more) goroutines which can penetrate the first `if` fighting with each other to initialise the Singleton itself, hence, the Lock/Unlock after the first `if`.

Reply:
So, I actually did that intentionally.
Here, the first `if` is to just provide non-locked reads to any future readers/goroutines calling this function because there's no point to Lock/Unlock once the Singleton is initialised as it's gonna be just GET THE SINGLETON afterwards. Hence, the first `if` isn't preceded with Locking it.
But in theory, before the Singleton is initialized, there can be two (or more) goroutines which can penetrate the first `if` fighting with each other to initialise the Singleton itself, hence, the Lock/Unlock after the first `if`.

Reply:
You are still allowing concurrent access to the same variable, which leads to undefined behavior!
You can't just declare some memory races safer than others, because of a gut feeling.

I highly recommend reading through the go memory model: https://go.dev/ref/mem, especially the sections on `Synchronization` and `Incorrect synchronization`.

Reply:
Ahh, I misunderstood Go's memory model to respect happens-before relationship from L54-L57 (in this case). 
Though to fight of that, we can wrap up the entire body of this function under an `RLock() and RUnlock()` but since https://github.com/openshift/reference-addon/pull/9/commits/d23969f5eb41305794a0d57e8e9be727ced15358 doesn't try to return any error during the initialization, I might as well use `once.Do` to do initialize the singleton safely.

Reply:
https://github.com/openshift/reference-addon/pull/9#:~:text=add%20concurrency%2Dsafe%20initialization%20of%20the%20singleton%20and%20some%20cleanup%E2%80%A6